sky/jobs/state.py

```diff
@@ -2587,3 +2638,139 @@ def set_controller_logs_cleaned(job_ids: List[int], logs_cleaned_at: float):
+            job-level event that applies to all tasks.
+        new_status: The new status being transitioned to. Can be a
+            ManagedJobStatus enum.
+        reason: A description of why the event occurred.
+        code: Optional error category code for failures.
+        timestamp: The timestamp of the event. If None, uses current time.
+    """
+    if timestamp is None:
+        timestamp = datetime.datetime.now()
+
+    status_value = new_status.value
+
+    assert _SQLALCHEMY_ENGINE_ASYNC is not None
+    async with sql_async.AsyncSession(_SQLALCHEMY_ENGINE_ASYNC) as session:
+        await session.execute(job_events_table.insert().values(
+            spot_job_id=job_id,
+            task_id=task_id,  # Can be None for job-level events
+            new_status=status_value,
+            code=code,
+            reason=reason,
+            timestamp=timestamp,
+        ))
+        await session.commit()
+
+
+@_init_db
+def get_job_events(job_id: int,
+                   task_id: Optional[int] = None,
+                   limit: Optional[int] = None) -> List[Dict[str, Any]]:
+    """Get task events for a managed job.
+
+    Args:
+        job_id: The spot_job_id of the managed job.
+        task_id: Optional task_id to filter by. If None, returns events
+            for all tasks. If specified, returns events for that task plus
+            job-level events (where task_id is None).
+        limit: Optional limit on number of events to return. If specified,
+            returns the most recent N events.
+
+    Returns:
+        List of event records, ordered by timestamp descending
+        (most recent first) if limit is specified, otherwise ascending.
+    """
+    assert _SQLALCHEMY_ENGINE is not None
+    with orm.Session(_SQLALCHEMY_ENGINE) as session:
+        query = sqlalchemy.select(
+            job_events_table.c.spot_job_id,
+            job_events_table.c.task_id,
+            job_events_table.c.new_status,
+            job_events_table.c.code,
+            job_events_table.c.reason,
+            job_events_table.c.timestamp,
+        ).where(job_events_table.c.spot_job_id == job_id)
+
+        if task_id is not None:
+            # Include events for the specific task AND job-level events
+            # (task_id is None)
+            query = query.where(
+                sqlalchemy.or_(job_events_table.c.task_id == task_id,
+                               job_events_table.c.task_id.is_(None)))
+
+        # Order by timestamp descending to get most recent first
+        query = query.order_by(job_events_table.c.timestamp.desc())
+
+        if limit is not None:
+            query = query.limit(limit)
+
+        rows = session.execute(query).fetchall()
+        return [{
+            'spot_job_id': row[0],
+            'task_id': row[1],
+            'new_status': ManagedJobStatus(row[2]),
+            'code': row[3],
+            'reason': row[4],
+            'timestamp': row[5],
+        } for row in rows]
```

Comment:
minor: should we keep this out of the session context to reduce the time in a session?